### PR TITLE
added quotation api

### PR DESCRIPTION
**Pull request by pierlucatonioloreply**
body used to test:

{
    "id": null,
    "idSign": null,
    "idPay": null,
    "contractFileId": null,
    "uid": null,
    "productUid": null,
    "productVersion": "v1",
    "proposalNumber": null,
    "number": null,
    "numberCompany": null,
    "status": null,
    "company": "axa",
    "name": "life",
    "nameDesc": "Wopta per te Gap",
    "startDate": "2023-05-12T15:37:14.009Z",
    "endDate": "2028-05-12T15:37:14.009Z",
    "creationDate": "0001-01-01T00:00:00.000Z",
    "updated": "0001-01-01T00:00:00.000Z",
    "payment": null,
    "paymentType": null,
    "paymentSplit": null,
    "isPay": null,
    "isAutoRenew": null,
    "isSign": null,
    "coverageType": null,
    "voucher": null,
    "channel": null,
    "covenant": null,
    "taxAmount": null,
    "priceNett": null,
    "priceGross": null,
    "priceNettMonthly": null,
    "priceGrossMonthly": null,
    "quoteQuestions": null,
    "contractor": {},
    "assets": [
        {
            "name": "gap",
            "address": "Via Carlo Vidua 17",
            "type": "vehicleType",
            "building": null,
            "person": {
                "residence": {
                    "locality": "VI"
                }
            },
            "enterprise": null,
            "isContractor": false,
            "guarantees": null,
            "vehicle": {
                "plate": "AA111AA",
                "manufacturer": "Opel",
                "registrationDate": "2023-05-12T15:37:14.009Z",
                "ownershipStatus": "si",
                "numberOfOwners": 1,
                "priceValue": 40000,
                "vehicleType": "car",
                "weight": "1302",
                "powerSupply": "benzina",
                "hasSatellite": false
            }
        }
    ],
    "attachments": [],
    "statements": [],
    "surveys": [],
    "offersPrices": null,
    "offerName": null,
    "codeCompany": null
}

### CODE REVIEW

File: quote/gap.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	"strings"
+	"time"
+
+	"github.com/go-gota/gota/dataframe"
+	lib "github.com/wopta/goworkspace/lib"
+	"github.com/wopta/goworkspace/models"
+)
+
+func GapFx(w http.ResponseWriter, r *http.Request) (string, interface{}, error) {
+	req := lib.ErrorByte(io.ReadAll(r.Body))
```
Comment (diogo-wopta) [inline quote/gap.go:22]: Add logging for handler start.
log.Println("[GapFx] Handler start")

File: quote/gap.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	var data models.Policy
+	defer r.Body.Close()
+	e := json.Unmarshal(req, &data)
+
+	authToken, err := models.GetAuthTokenFromIdToken(r.Header.Get("Authorization"))
+	lib.CheckError(err)
+	res, e := Gap(authToken.Role, data)
```
Comment (diogo-wopta) [inline quote/gap.go]: Why are we passing the user role if it's not being used?

File: quote/gap.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/go-gota/gota/dataframe"
+	lib "github.com/wopta/goworkspace/lib"
+	"github.com/wopta/goworkspace/models"
+)
+
+func GapFx(w http.ResponseWriter, r *http.Request) (string, interface{}, error) {
+	req := lib.ErrorByte(io.ReadAll(r.Body))
+	var data models.Policy
+	defer r.Body.Close()
+	e := json.Unmarshal(req, &data)
```
Comment (diogo-wopta) [inline quote/gap.go:26]: Handle error with lib.CheckError(e)

File: quote/gap.go (hunk 2 above)
Comment (diogo-wopta) [inline quote/gap.go]: We are not handling the error, but seeing the Gap function the error is never set internally.

File: quote/gap.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	var residenceCode string = data.Assets[len(data.Assets)-1].Person.Residence.Locality
+	var vehicleValue int64 = data.Assets[len(data.Assets)-1].Vehicle.PriceValue
+
+	//get the residence area
+	residenceArea := getResidentArea(provincesMatrixDF, residenceCode)
+
+	//get the duration
+	duration := getDuration(data.StartDate, data.EndDate)
+
+	fmt.Printf("valore base e' %d\n", duration)
```
Comment (diogo-wopta) [inline quote/gap.go]: use the log library with english messages

File: quote/gap.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+	authToken, err := models.GetAuthTokenFromIdToken(r.Header.Get("Authorization"))
+	lib.CheckError(err)
+	res, e := Gap(authToken.Role, data)
+	s, e := json.Marshal(res)
+	return string(s), nil, e
+
+}
+
+func Gap(role string, data models.Policy) (models.Policy, error) {
```
Comment (diogo-wopta) [inline quote/gap.go]: We could pass the policy (data) as a pointer as to modify incrementally the same object (this is also valid for the function setOfferPrices)

File: quote/gap.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	data.OffersPrices = make(map[string]map[string]*models.Price)
+
+	data.OffersPrices["base"] = map[string]*models.Price{
+		"singleInstallment": {
+			Net:      baseGapMultiplierFloat * float64(vehicleValue),
+			Tax:      0.0,
+			Gross:    0.0,
+			Delta:    0.0,
+			Discount: 0.0,
+		},
```
Comment (diogo-wopta) [inline quote/gap.go]: we could extract the construction of an offer price into a function (the only difference is the multiplier) that internally do all the math for calculating Net, Tax and Gross values

File: quote/gap.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	return data
+}
+
+func getResidentArea(provincesMatrixDF dataframe.DataFrame, residenceCode string) string {
+	for _, row := range provincesMatrixDF.Records() {
+		//fmt.Println(row)
```
Comment (diogo-wopta) [inline quote/gap.go]: remove commented code

File: quote/gap.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+	//get the:
+	// - vehicle owner residence
+	// - vechicle value
+	var residenceCode string = data.Assets[len(data.Assets)-1].Person.Residence.Locality
+	var vehicleValue int64 = data.Assets[len(data.Assets)-1].Vehicle.PriceValue
+
+	//get the residence area
+	residenceArea := getResidentArea(provincesMatrixDF, residenceCode)
+
```
Comment (Yousef-Hammar) [inline quote/gap.go:57]: add check if residence area is equal to empty string and call checkerror

File: quote/gap.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	completeMatrixDF := lib.CsvToDataframe(completeMatrix)
+	provincesMatrixDF := lib.CsvToDataframe(provincesMatrix)
+
+	//get the:
+	// - vehicle owner residence
+	// - vechicle value
+	var residenceCode string = data.Assets[len(data.Assets)-1].Person.Residence.Locality
+	var vehicleValue int64 = data.Assets[len(data.Assets)-1].Vehicle.PriceValue
+
+	//get the residence area
```
Comment (Yousef-Hammar) [inline quote/gap.go:55]: add log saying get area using standard format (e.g., [GAP] )

File: quote/gap.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+	log.Printf("base value e': %d\n", duration)
+
+	//get the base and complete multipliers
+	baseGapMultiplierFloat, completeGapMultiplierFloat := getGapMultipliers(baseMatrixDF, completeMatrixDF, duration, residenceArea)
+
+	//get the matrix area row
+	log.Printf("valore base e' %f\n", baseGapMultiplierFloat)
+	log.Printf("valore completo e' %f\n", completeGapMultiplierFloat)
+	log.Printf("valore base veicolo*tax e' %f\n", baseGapMultiplierFloat*float64(vehicleValue))
+	log.Printf("valore completo veicolo*tax e' %f\n", completeGapMultiplierFloat*float64(vehicleValue))
```
Comment (Yousef-Hammar) [inline quote/gap.go:70]: remove these two logs

File: quote/gap.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+	duration := getDuration(data.StartDate, data.EndDate)
+
+	log.Printf("base value e': %d\n", duration)
+
+	//get the base and complete multipliers
+	baseGapMultiplierFloat, completeGapMultiplierFloat := getGapMultipliers(baseMatrixDF, completeMatrixDF, duration, residenceArea)
+
+	//get the matrix area row
+	log.Printf("valore base e' %f\n", baseGapMultiplierFloat)
+	log.Printf("valore completo e' %f\n", completeGapMultiplierFloat)
```
Comment (Yousef-Hammar) [inline quote/gap.go:68]: change message with baseMultiplier and completeMultiplier

File: quote/gap.go [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+		return gapMultipierFloat * float64(vehicleValue)
+	case "tax":
+		return taxValue
+	case "gross":
+		return (gapMultipierFloat * float64(vehicleValue)) + taxValue
+	default:
+		return 0
+	}
+
+}
```
Comment (Yousef-Hammar) [inline quote/gap.go:126]: remove this

File: quote/gap.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+	//get the residence area
+	residenceArea := getResidentArea(provincesMatrixDF, residenceCode)
+
+	//get the duration
+	duration := getDuration(data.StartDate, data.EndDate)
+
+	log.Printf("base value e': %d\n", duration)
+
+	//get the base and complete multipliers
+	baseGapMultiplierFloat, completeGapMultiplierFloat := getGapMultipliers(baseMatrixDF, completeMatrixDF, duration, residenceArea)
```
Comment (Yousef-Hammar) [inline quote/gap.go:64]: insert multipliers in a map[string]float64 where the key if the offerName (e.g., base/complete) and the value will be the multiplier.
Use this map inside the function setOfferPrices

File: quote/gap.go [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+	for offerKey, offerValue := range offersPrices {
+		for paymentKey := range offerValue {
+			offersPrices[offerKey][paymentKey].Net = lib.RoundFloat(offersPrices[offerKey][paymentKey].Net, 2)
+			offersPrices[offerKey][paymentKey].Tax = lib.RoundFloat(offersPrices[offerKey][paymentKey].Tax, 2)
+			offersPrices[offerKey][paymentKey].Gross = lib.RoundFloat(offersPrices[offerKey][paymentKey].Gross, 2)
+		}
+	}
+}
+
+func setOffersPrices(data *models.Policy, vehicleValue int64, baseGapMultiplierFloat float64, completeGapMultiplierFloat float64) {
```
Comment (Yousef-Hammar) [inline quote/gap.go:88]: input to this function (data *models.Policy, vehicleValue int64, multipliers map[string]float64)
use a for loop to set the prices for each offer

**Review by Yousef-Hammar — CHANGES_REQUESTED** [review]

**Pull request closed without merging**